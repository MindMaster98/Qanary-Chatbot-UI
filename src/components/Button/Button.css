.button {
    box-shadow: 0px 2px 6px 0px rgba(0, 0, 0, 0.24);
    border-radius: 7px;
    min-height: fit-content;
    min-width: fit-content;
}