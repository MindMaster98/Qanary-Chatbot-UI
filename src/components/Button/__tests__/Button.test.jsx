import renderer from "react-test-renderer";
import { act } from "@testing-library/react";
import ReactDOM from "react-dom";
import Button from "../Button";

// Snapshot tests

it("renders correctly with no props", () => {
  const tree = renderer.create(<Button />).toJSON();
  expect(tree).toMatchSnapshot();
});

it("renders correctly with only text", () => {
  const tree = renderer.create(<Button text="test" />).toJSON();
  expect(tree).toMatchSnapshot();
});

it("renders correctly with inline style", () => {
  const tree = renderer.create(<Button style={{ color: "red" }} />).toJSON();
  expect(tree).toMatchSnapshot();
});

it("renders correctly with only undefined props", () => {
  const tree = renderer
    .create(
      <Button
        text={undefined}
        className={undefined}
        style={undefined}
        onClick={undefined}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("renders correctly with only null props", () => {
  const tree = renderer
    .create(<Button text={null} className={null} style={null} onClick={null} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

// Blackbox tests
// You cannot dispatch events using test renderer, so I need to actually add the element to the DOM 
// and dispatch an actual event. Calling the function directly would oppose the idea of a blackbox.
// Why not enzyme? Because enzyme doesn't officially support React 17.

let container;

beforeEach(() => {
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  document.body.removeChild(container);
  container = null;
});

it("calls the onClick function when clicked", () => {
  const mockFn = jest.fn();

  act(() => {
    ReactDOM.render(<Button onClick={mockFn} />, container);
  });

  const btn = document.querySelector("button");
  btn.dispatchEvent(new MouseEvent("click", { bubbles: true }));

  expect(mockFn).toHaveBeenCalledTimes(1);
});

it("does not call the onClick function when not clicked", () => {
  const mockFn = jest.fn();

  act(() => {
    ReactDOM.render(<Button onClick={mockFn} />, container);
  });

  expect(mockFn).toHaveBeenCalledTimes(0);
});
