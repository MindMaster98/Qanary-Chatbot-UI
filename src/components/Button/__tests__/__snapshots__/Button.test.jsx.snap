// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with inline style 1`] = `
<button
  className="button "
  style={
    Object {
      "color": "red",
    }
  }
>
  
</button>
`;

exports[`renders correctly with no props 1`] = `
<button
  className="button "
  style={Object {}}
>
  
</button>
`;

exports[`renders correctly with only null props 1`] = `
<button
  className="button "
  style={Object {}}
>
  
</button>
`;

exports[`renders correctly with only text 1`] = `
<button
  className="button "
  style={Object {}}
>
  test
</button>
`;

exports[`renders correctly with only undefined props 1`] = `
<button
  className="button "
  style={Object {}}
>
  
</button>
`;
